File: client-flutter/lib/services/appointment_service.dart
import '../models/appointment.dart';
import '../models/calendar_data.dart';
import 'api_client.dart';

class AppointmentService {
  Future<StartData> getStartData(String clientId) async {
    final data = await apiClient.get('api/client/start/$clientId');
    if (data == null || data is! Map<String, dynamic>) {
      throw ApiException(500, 'Ungültige Startdaten');
    }
    return StartData.fromJson(data);
  }

  Future<CalendarData> getCalendarData(String clientId) async {
    final data = await apiClient.get('api/client/calendar/$clientId');
    if (data == null || data is! Map<String, dynamic>) {
      throw ApiException(500, 'Ungültige Kalenderdaten');
    }
    return CalendarData.fromJson(data);
  }

  Future<Map<String, dynamic>> bookAppointment(
    String clientId, {
    required String trainerId,
    required String trainingTypeId,
    required String date,
    required String starttime,
    String? locationId,
    required int duration,
  }) async {
    final body = <String, dynamic>{
      'trainer_id': int.tryParse(trainerId) ?? trainerId,
      'training_type_id': int.tryParse(trainingTypeId) ?? trainingTypeId,
      'date': date,
      'starttime': starttime,
      'duration': duration,
    };
    if (locationId != null && locationId.isNotEmpty) {
      body['location_id'] = int.tryParse(locationId) ?? locationId;
    }
    final data = await apiClient.post(
      'api/client/appointment/$clientId',
      body: body,
    );
    return data is Map<String, dynamic> ? data : {};
  }

  /// Returns the full response map on success (includes `creditRefunded`), null on failure.
  Future<Map<String, dynamic>?> cancelAppointment(String clientId, String appointmentId) async {
    final data = await apiClient
        .delete('api/client/appointment/$clientId/$appointmentId');
    if (data != null && data['success'] == true) return data;
    return null;
  }
}

File: client-ios/SihlClient/Services/AppointmentService.swift
import Foundation

/// Pendant zu `services/appointment_service.dart`.
/// Alle Methoden sind stateless; Instanz bei Bedarf wegwerfen oder wiederverwenden.
struct AppointmentService {

    func getStartData(clientId: String) async throws -> StartData {
        let data = try await APIClient.shared.get("api/client/start/\(clientId)")
        guard let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            throw APIError(statusCode: 500, message: "Ungültige Startdaten")
        }
        return StartData(json: json)
    }

    func bookAppointment(
        clientId: String,
        trainerId: String,
        trainingTypeId: String,
        date: String,
        starttime: String,
        locationId: String?,
        duration: Int
    ) async throws {
        var body: [String: Any] = [
            "trainer_id":        Int(trainerId)       as Any? ?? trainerId,
            "training_type_id":  Int(trainingTypeId)  as Any? ?? trainingTypeId,
            "date":      date,
            "starttime": starttime,
            "duration":  duration,
        ]
        if let loc = locationId, !loc.isEmpty {
            body["location_id"] = Int(loc) as Any? ?? loc
        }
        _ = try await APIClient.shared.post("api/client/appointment/\(clientId)", body: body)
    }

    /// Gibt den vollen Response-Body zurück (enthält `creditRefunded`), oder `nil` wenn
    /// `success != true` — analog zur Flutter-Implementierung.
    func cancelAppointment(clientId: String, appointmentId: String) async throws -> [String: Any]? {
        let data = try await APIClient.shared.delete(
            "api/client/appointment/\(clientId)/\(appointmentId)"
        )
        guard
            let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
            json["success"] as? Bool == true
        else { return nil }
        return json
    }
}
